experiments/experiment.go

```diff
@@ -132,9 +144,56 @@ func (ec *experimentContext) reconcileTemplate(template v1alpha1.TemplateSpec) {
 			}
 		}
 	} else {
+		if desiredReplicaCount == 0 {
+			// Add scaleDownDelay if necessary
+			rsIsUpdated, err := ec.addScaleDownDelay(rs)
+			if err != nil {
+				ec.log.Warnf("Unable to add scaleDownDelay label on rs '%s'", rs.Name)
+			} else {
+				if rsIsUpdated {
+					ctx := context.TODO()
+					modifiedRS, err := ec.kubeclientset.AppsV1().ReplicaSets(ec.ex.Namespace).Get(ctx, rs.Name, metav1.GetOptions{})
+					if err != nil {
+						ec.log.Warnf("Unable to get rs '%s' with added scaleDownDelay", rs.Name)
+					} else {
+						rs = modifiedRS
+						ec.templateRSs[template.Name] = modifiedRS
+					}
+				}
+			}
+
+		}
 		// Replicaset exists. We ensure it is scaled properly based on termination, or changed replica count
 		if *rs.Spec.Replicas != desiredReplicaCount {
-			ec.scaleReplicaSetAndRecordEvent(rs, desiredReplicaCount)
+			if desiredReplicaCount == 0 {
+				// Add delay before scaling
+				if scaleDownAtStr, ok := rs.Annotations[v1alpha1.DefaultReplicaSetScaleDownDeadlineAnnotationKey]; ok {
+					scaleDownAtTime, err := time.Parse(time.RFC3339, scaleDownAtStr)
+					if err != nil {
+						ec.log.Warnf("Unable to read scaleDownAt label on rs '%s'", rs.Name)
+					}
+					now := metav1.Now()
+					scaleDownAt := metav1.NewTime(scaleDownAtTime)
+					if scaleDownAt.After(now.Time) {
+						ec.log.Infof("RS '%s' has not reached the scaleDownTime", rs.Name)
+						remainingTime := scaleDownAt.Sub(now.Time)
+						if remainingTime < ec.resyncPeriod {
```

Comment:
Instead of copying and pasting this code which is now duplicated in rollout controller, the scaledown helper logic should be moved to its own set of helper libraries that can be reused between the  rollout and experiment controller.